cql3/restrictions/restriction.hh

```diff
@@ -58,6 +61,59 @@ namespace restrictions {
 struct allow_local_index_tag {};
 using allow_local_index = bool_class<allow_local_index_tag>;
 
+/// Work-in-progress new restriction representation.  When complete, it will replace the old one.
+namespace wip {
+
+// The new representation exposes its member data publicly.  Operations on restrictions are
+// performed by free functions that take restrictions as parameters and use the visitor pattern to
+// specialize code for different kinds of restrictions.
+//
+// The most interesting class is wip::binary_operator below, which can represent both multi- and
+// single-column restrictions.  Instead of the old merge_with mechanism for combining restrictions,
+// we will simply add them into a conjunction expression that will be processed using visitors.
+
+// Forward declare these, so std::variant can take them as arguments:
+struct binary_operator;
+struct conjunction;
+struct subscript;
+struct token;
+
+/// A restriction expression -- union of all possible restriction types.
+using expression = std::variant<binary_operator, conjunction>;
+
+/// Identifier(s) on LHS of an operator restriction.
+using idents = std::vector<const column_definition*>;
+
+/// LHS of an operator restriction.
+using lhs_t = std::variant<idents, token, subscript>;
```

Comment:
Let's plan to eliminate this, at least in the context of expressions. For now it makes sense because cql gives special treatment to the lhs and rhs.

Reply:
Done, in a [future version](https://github.com/dekimir/scylla/commit/6ca09dc6b97b969360bd7ae0ea3a22870d054c4f#diff-2b4396b4d6732711f7747468eb6614f2L94) that hasn't landed here yet.

Reply:
That future version has now landed.